pep/004-labels-and-namespaces.md

```diff
@@ -0,0 +1,386 @@
+
+#### Bundle Labels
+
+A new section in porter.yaml stores labels:
+
+```yaml
+name: mybun
+version: v0.1.0
+
+labels:
+  app: myapp
+  io.cnab/appVersion: v1.2.3
+  internal: "true"
+```
+
+I am [proposing labels to the CNAB Spec][cnab-labels-proposal] as a first class value on the bundle, reserving the `io.cnab/` prefix, and defining common labels that tools can use to filter bundles:
+
+* io.cnab/appVersion - The version of the application installed by the bundle.
+
+If accepted labels would be stored in the bundle.json like so:
+
+```json
+{
+  "name": "mybun",
+  "version": "v0.1.0",
+  "labels": {
+    "app": "myapp",
+    "io.cnab/appVersion": "v1.2.3",
+    "internal": "true"
+  }
+}
+```
+
+[cnab-labels-proposal]: https://github.com/cnabio/cnab-spec/pull/409
+
+#### Installation Labels
+
+When a bundle is installed, the labels defined on the bundle are copied and used as labels on the installation.
+Users can apply additional labels using the --label flag, which can be repeated, has a short flag of `-l` and uses the following format: KEY=VALUE.
+
+Labels can be applied to an installation when the bundle is installed:
+```
+porter install wordpress --reference getporter/wordpress:v1.0 --label owner=carolyn
+```
+
+Labels can be applied after an installation is created using the `porter installation label` command:
+
+```
+$ porter installation label set wordpress --label env=dev --label internal=true
+# add (or updates) env=dev and internal=true to the wordpress installation
+
+$ porter installation label delete wordpress --label env
+# Remove the env label from the wordpress installation
+```
+
+🚨 The CNAB Spec does not cover storing installation data, and we should push for having a document representation of installations be added to the spec.
+
+```json
+{
+  "name": "wordpress",
+  "created": "TIMESTAMP",
+  "bundleRepository": "getporter/wordpress",
+  "version": "0.1.0",
+  "labels": {
+    "app": "wordpress",
+    "io.cnab/appVersion": "v1.2.3",
+    "owner": "carolyn"
+  },
+  "custom": {}
+}
+```
+
+Users can filter by labels when listing installations:
+
+```console
+$ porter installations list --label owner=carolyn
+# Filter installations by the label owner=carolyn
+```
+
+#### Claim Labels
+
+In Porter a claim is represented as a "run".
+Each time porter install/upgrade/invoke/uninstall is executed, that generates a claim (except when the action is stateless such as io.cnab.status).
+The user-defined labels provided when the bundle is installed are applied to both the installation and the claim.
+
+```
+porter install wordpress --reference getporter/wordpress:v1.0 --label owner=carolyn
+```
+
+**claim**
+```json
+{
+  "id": "CLAIM_ID",
+  "installation": "wordpress",
+  "action": "install",
+  "labels": {
+    "owner": "carolyn"
+  }
+}
+```
+
+During upgrade, invoke and uninstall the user-defined labels are applied to the claim only:
```

Comment:
Curious around the design decision here around _not_ updating the installation with the user-defined label(s) for these actions?  (And we might consider adding that decision to the doc for context...)

Reply:
I think I'm trying to put too much into the install command here because I wanted to avoid either of the situations below

**applying installation labels separate from the installation creation**
```
porter install wordpress
porter installation label set owner=carolynvs
```

**adding weird one-off flags to install**
```
porter install wordpress --installation-label owner=carolynvs
```

But what I suggested in the proposal is really inconsistent (sometimes --label applies to an installation label and sometimes a claim label)...

I guess out of all these options, I would go with "you only set installation labels using the porter installation label command" and not try to pack a unique behavior into install.

Reply:
Yea, the inconsistency might feel weird.  I'm still unclear on why we wouldn't want to, say, `porter upgrade wordpress --label owner=carolynvs` to update the installation as well?  If I would be able to run `porter upgrade wordpress && porter installation label set wordpress owner=carolynvs`, why not allow for the inline `--label` on any action to take effect (modify the installation as well as the claim)?

Reply:
I guess I'm getting hung up on being able to set labels on the claim and the installation independently, you know?

In my head I was imagining that it could be useful to put a label _just_ on the claim, to communicate something about that particular run of the bundle. For example, the operator could add the label "porter.sh/operator-revision=REVISION_FROM_KUBERNETES_THAT_TRIGGERED_THE_RUN". I really wouldn't want that stored on the installation as well, it should just be on the claim.

Reply:
I'm going to redo this with `porter install --label-installation KEY=VALUE`. I think that's the most clear out of all the options. Then `--label` on all of the actions always maps to the claim labels. After the installation exists, people can use `porter installation label set` to manage the labels.